### DISCUSSION TITLE
Add a new hook for before and after a connection happens

### DISCUSSION THREAD

**Discussion by pfreixes** [redis/go-redis#1786] [Ideas]
The idea is extending the current Hook interface [1] for adding two new available functions, for example `BeforeConnect` and `AfterConnect`, that would be called before a new connection is asked and before.

These two new hooks could be used for extending observability that the client side, providing a reliable metric on the cost of creating new connections in terms of latency.

Connection errors could be also provided as a parameter of the `AfterConnect` function.

I could take care of making the proper PR with the implementation.

[1] https://github.com/go-redis/redis/blob/6e4eb2e3acad9575fb1e2c8690a3c1decf2e59e5/redis.go#L24

**Comment by vmihailenco** [comment]
Sounds good to me :+1: 

I believe this new hook should be optional:

```go
type ConnectEvent struct {
    Err error
}

type ConnectHook interface {
    BeforeConnect(ctx context.Context) context.Context
    AfterConnect(ctx context.Context, event ConnectEvent)
}

type fullHook interface {
    Hook
    ConnectHook
}

type hooks struct {
	hooks []fullHook
}
```

Then in `Redis.AddHook` you can check if the provided `hook` implements the `fullHook` interface and upgrade it to `fullHook` if necessary (just use a wrapper with empty `*Connect` methods).

Does that make sense?

**Reply by pfreixes** [reply]
Ill work on the PR

**Reply by pfreixes** [reply]
Im a bit lost with your suggestion. Shouldn´t the `FullHook` be public? as a way for providing by the caller the full set of hooks? Considering that would be only two options valid here, the current Hook or the FullHook.

This will force us to make a casting behind the scenes (when `AddHook` is called) but still keeping the original `Hook` type, which not sure how easy is gonna be for the user to understand that behind the scenes we are making the casting for also supporting the `FullHook`. Would it make sense to have a new `AddFullHook`?

Regarding the wrapper, and considering that that `FullHook` is an interface, this will force us to have its own internal object that acts as a proxy for the none optional methods (`AfterProccess`, `BeforeProcess`, ...) and makes a noop when the connection  hooks are called. isn´t it? or is there a way for providing `nil` (no implementation) for the noop methods? Should we delegate this responsibility to the user, if its possible, and ask the user to provide a `FullHook` which has selectively implemented only the methods that its interested and setting as `nil` the other ones? 

If eventually we end up adding more hooks, how this would work? so making optional more methods?

**Reply by vmihailenco** [reply]
What do you think about using a Dialer option instead:

```go
redis.NewClient(&redis.Options{
    Dialer: func(ctx context.Context, network, addr string) (net.Conn, error) {
        fmt.Println("before connect")
        conn, err := net.Dial(network, addr)
        fmt.Println("after connect")
        return conn, err
    }
})
```

**Reply by pfreixes** [reply]
Not really fan, but if we can not find a proper way for adding the hooks could be an alternative.

What would be you red flags or no go in terms of design? breaking the current contract of hooks? Ill would like to keep stick to the hooks pattern.

For example what about adding the `BeforeConnect` and `AfterConnect` as first citizen hooks, none optionals. Requesting to the user to update the hooks interface and providing noop functions if the user is not interested on tracing the connection.

**Reply by vmihailenco** [reply]
This is how we could keep the existing Hook interface as is and add ConnectHook - https://github.com/go-redis/redis/pull/1791

But now I think that we should recommend using `Options.Dialer` and not add such a hook. Not many people need such functionality so cumbersome API should not be a problem.

**Reply by pfreixes** [reply]
Oks, seems that this is pretty much done, do you need any help?

**Reply by pfreixes** [reply]
Sorry for the late response here, Ive been working on a delta version of your proposal which moves all of the pool hook objects under the pool module, but still using as entrypoint for adding a hook what you proposed. Here a draft on how this could look like https://github.com/pfreixes/redis/commit/1a01c44b2a3baf709b8039f732d4ea65aa0efc07

For the sake of calling the hooks if and only if a new connection is created, which is responsibility of the Pool, and for avoid doing the wiring at `Get` method level, just decided that maybe would be worth it to add the hooks at Pool type level, and then leave to it the responsibility to call it when its necessary.

**Comment by monkey92t** [comment]
It sounds good, but I can hardly imagine what benefits this hook can bring? Why should we add it?
If you just observe the cost of creating a connection, I think it’s easier to use Options.Dialer and Options.OnConnect.

**Reply by pfreixes** [reply]
Hey @monkey92t thanks for the questions.

I guess its about separating concerns and using the right interfaces, I agree with you that with overriding the `Options.Dialer` you could implement the proper observability for measuring the time needed for creating a connection, but not sure if originally this was the main reason for providing a way for overriding the dialer, maybe you could help me to understand what was the reason on adding this option.

In any case this feature wants to solve the lack of having observability for when new connections are created by providing the same "interface" of already existing hooks, but providing its own hooks. This makes adding observability almost friction less for third party which they only take care of implementing their own logic that needs to be executed before and after, without having to bother on doing some extra responsibilities (like calling the original dialer) 

Also remember that hooks allow you to bind as many hooks as you want, by using the `Options.Dialer` solution this will not work.

**Reply by monkey92t** [reply]
I understand the function of hooks, and I am wondering whether it is necessary to add it to the connection pool. go-redis uses long connections and will not create new connections under normal circumstances.

**Reply by monkey92t** [reply]
Sometimes it runs for a few weeks and only creates a few new connections when it starts running.

**Reply by monkey92t** [reply]
It is not difficult for us to add it, I wonder if someone really needs such a feature.
The reason for adding the `options.Dialer` option is to create a network connection as an option. Some people may not use the default `net.Dial` method but customize the connection function. This is a common practice.

**Reply by pfreixes** [reply]
Measuring latency is an extension in terms of observability to the current stats [1]. Connections can be created at any moment under many different circumstances, pool creation, spike of traffic, connection recycling due to errors (timeouts), ....

Latency of the connection creation is not rare in the industry 

[1] https://github.com/go-redis/redis/blob/master/internal/pool/pool.go#L36

**Reply by pfreixes** [reply]
just for the sake of curiosity

> Some people may not use the default net.Dial method but customize the connection function

Any chance to have some examples of the kind of customisation that people use to make by providing alternative dial functions?

**Reply by pfreixes** [reply]
No hard feelings for me, but a proposal here [1] in case you are still interested on adding the hooks instead of using the override of the `dial` command.

Its still missing a solution for the cloned clients since current implementation won't create a new slice of connection pool hooks for each cloned client, any recommendation will be more than welcomed. 

If there is no a good solution but you are still believe that this solution is cleaner and could get eventually merged we might think on moving the connection pool hooks under the Redis type, with all of the consequences that it would have (having to provide the hooks as a parameter for the `pool.Get` method)

[1] https://github.com/go-redis/redis/pull/1803

**Reply by vmihailenco** [reply]
>Latency of the connection creation is not rare in the industry

That was my impression as well when we created tracing instrumentation for OpenTelemetry. We used to measure dial, read, and write timings.

But in practice it was not very useful. Dials happen so rarely that you can't even draw a nice chart for it. Reads/writes happen regularly, but all the time is spent waiting/reading data from sockets so in the end it is just (expensive to process) noise.

So I think that we don't need a separate hook given that you can wrap `Dialer` to get the timings. And we are going to add proper metrics to go-redis using OpenTelemetry Metrics (in 1-2 months). Connection-related metrics will be added first. I think that should cover 99% needs and the rest can be achieved with (clumsy) dialer-wrapping API.

**Reply by MakingL** [reply]
I now want to test the latency of my connection establishment. I will initiate multiple connections concurrently to test the connection processing capability of my server, so I need to look at all connection establishment latency. Can you give some suggestions to implement this function through `Dialer`? @vmihailenco
